src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -0,0 +1,244 @@
+                  (Ps.hash ancestor) proof ~ancestor_length
+              with
+              | Ok () ->
+                  t.best_with_root <- {state= candidate; root= ancestor} ;
+                  let candidate_body_hash = Ps.Body.hash (Ps.body candidate) in
+                  let candidate_hash =
+                    Protocol_state.hash ~hash_body:Fn.id
+                      {body= candidate_body_hash; previous_state_hash}
+                  in
+                  Ancestor.Prover.add t.ancestor_prover ~hash:candidate_hash
+                    ~prev_hash:previous_state_hash
+                    ~length:
+                      Ps.(Consensus_state.length (consensus_state candidate))
+                    ~body_hash:candidate_body_hash ;
+                  let ledger_hash =
+                    Consensus.Mechanism.(
+                      Protocol_state.blockchain_state ancestor
+                      |> Blockchain_state.ledger_hash
+                      |> Frozen_ledger_hash.to_ledger_hash)
+                  in
+                  ( match
+                      Syncable_ledger.new_goal t.syncable_ledger ledger_hash
+                    with
+                  | `Ignore -> ()
+                  | `Continue ->
+                      Ledger_hash.Table.set ledger_hash_table ~key:ledger_hash
+                        ~data:candidate_hash ) ;
+                  Deferred.unit
+              | Error e -> received_bad_proof t e ; get_root () )
+      in
+      get_root ()
+
+  let should_bootstrap root_state new_state =
+    let new_length = length new_state in
+    let root_length = length root_state in
+    new_length - root_length
+    > (2 * Transition_frontier.max_length) + Consensus.Mechanism.network_delay
+
+  let setup_bootstrap ~ledger_hash_table ~toggle_pipes ~frontier t
+      (transition, tm) =
+    toggle_pipes () ;
+    Transition_frontier.clear_paths frontier ;
+    don't_wait_for (on_transition t ~ledger_hash_table (transition, tm)) ;
+    let%map tree = Syncable_ledger.valid_tree t.syncable_ledger in
+    let root_hash = Ledger.merkle_root tree in
+    let state_hash = Hashtbl.find_exn ledger_hash_table root_hash in
+    Transition_frontier.rebuild frontier tree state_hash ;
+    Hashtbl.clear ledger_hash_table ;
+    toggle_pipes ()
+
+  let create ~parent_log ~frontier ~network ~ancestor_prover =
+    let logger = Logger.child parent_log __MODULE__ in
+    let initial_breadcrumb = Transition_frontier.root frontier in
+    let initial_root_state =
+      initial_breadcrumb |> Transition_frontier.Breadcrumb.transition_with_hash
+      |> With_hash.data |> External_transition.protocol_state
+    in
+    let ledger =
+      Transition_frontier.Breadcrumb.staged_ledger initial_breadcrumb
+      |> Staged_ledger.ledger
+    in
+    { network
+    ; logger
+    ; ancestor_prover
+    ; best_with_root= {state= initial_root_state; root= initial_root_state}
+    ; syncable_ledger= Syncable_ledger.create ledger ~parent_log:logger }
```

Comment:
I don't understand the relationship between `create` and `setup_bootstrap`, and `run`. Can these be made clearer?